File: frontend/lib/models/store.dart
class Store {
  final String id;
  final String name;
  final String? logoUrl;
  final Map<String, dynamic>? departmentOrder;
  final DateTime createdAt;
  final DateTime? updatedAt;

  Store({
    required this.id,
    required this.name,
    this.logoUrl,
    this.departmentOrder,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Store.fromJson(Map<String, dynamic> json) {
    return Store(
      id: json['id'] as String,
      name: json['name'] as String,
      logoUrl: json['logo_url'] as String?,
      departmentOrder: json['department_order'] as Map<String, dynamic>?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: json['updated_at'] != null ? DateTime.parse(json['updated_at'] as String) : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'logo_url': logoUrl,
        'department_order': departmentOrder,
        'created_at': createdAt.toIso8601String(),
        'updated_at': updatedAt?.toIso8601String(),
      };
}

class StoreDepartment {
  final String id;
  final String storeId;
  final String name;
  final int sortOrder;

  StoreDepartment({
    required this.id,
    required this.storeId,
    required this.name,
    required this.sortOrder,
  });

  factory StoreDepartment.fromJson(Map<String, dynamic> json) {
    return StoreDepartment(
      id: json['id'] as String,
      storeId: json['store_id'] as String,
      name: json['name'] as String,
      sortOrder: json['sort_order'] as int? ?? 0,
    );
  }
}

File: backend/app/models/store.py
"""Modele ORM sklepów i działów sklepowych."""

from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import JSON, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.session import Base


class Store(Base):
    """Sklep (np. Lidl, Biedronka, Dino)."""

    __tablename__ = "stores"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=func.gen_random_uuid(),
    )
    name: Mapped[str] = mapped_column(
        String(100), unique=True, nullable=False
    )
    logo_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    department_order: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), onupdate=func.now(), nullable=True
    )

    # ── Relacje ──────────────────────────────────────────────────
    departments: Mapped[list[StoreDepartment]] = relationship(
        "StoreDepartment", back_populates="store", cascade="all, delete-orphan"
    )
    store_products: Mapped[list["StoreProduct"]] = relationship(
        "StoreProduct", back_populates="store", cascade="all, delete-orphan"
    )

    def __repr__(self) -> str:
        return f"<Store {self.name!r}>"


class StoreDepartment(Base):
    """Dział w sklepie (np. 'Warzywa i owoce', 'Nabiał')."""

    __tablename__ = "store_departments"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=func.gen_random_uuid(),
    )
    store_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("stores.id", ondelete="CASCADE"),
        nullable=False,
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    sort_order: Mapped[int] = mapped_column(Integer, nullable=False, default=0)

    # ── Relacje ──────────────────────────────────────────────────
    store: Mapped[Store] = relationship("Store", back_populates="departments")
    store_products: Mapped[list["StoreProduct"]] = relationship(
        "StoreProduct", back_populates="department"
    )

    def __repr__(self) -> str:
        return f"<StoreDepartment {self.name!r} @ {self.store_id}>"
